alf/utils/video_recorder.py

```diff
@@ -179,23 +221,199 @@ def capture_frame(self, pred_info=None):
                         "be plotted when rendering videos.")
 
             self.last_frame = frame
-            if self.ansi_mode:
-                self._encode_ansi_frame(frame)
+            if defer_mode:
+                self._frame_buffer.append(frame)
+                self._observation_buffer.append(time_step.observation)
+                self._reward_buffer.append(time_step.reward)
+                self._action_buffer.append(policy_step.output)
+                self._encode_with_future_info(
+                    info_func=info_func, encode_all=time_step.is_last())
             else:
-                self._encode_image_frame(frame)
+                if self.ansi_mode:
+                    self._encode_ansi_frame(frame)
+                else:
+                    self._encode_image_frame(frame)
 
             assert not self.broken, (
                 "The output file is broken! Check warning messages.")
 
-    def _plot_prob_curve(self, name, probs, xticks=None):
+    def _encode_with_future_info(self, info_func=None, encode_all=False):
+        """Record future information and encode with the recoder.
+        This function extracts some information of ``future_steps`` into
+        the future, based on the input observations/actions/rewards.
+        By default, the reward and actions from the current time step
+        to the next ``future_steps`` will be displayed for each frame.
+        User can use ``info_func`` to add customized predictive quantities
+        to be shown in the video frames.
+
+        Args:
+            info_func (None|callable): a callable for calculating some customized
+                information (e.g. predicted future reward) based on the observation
+                at each time step and action sequences from the current time step
+                to the next ``future_steps`` steps (if available). It is called
+                as ``pred_info=info_func(current_observation, action_sequences)``.
+                Currently only support displaying scalar predictive information
+                returned from info_func.
+            encode_all (bool): whether to encode all the steps in the episode
```

Comment:
encode_all is not necessary because after self._future_steps < episode_buffer_length is satisfied, it will always be satisfied till the end.